### ISSUE TITLE
Thumbnails being downloaded with black bars added to top/bottom

### ISSUE THREAD

**Issue by ngInit** [meeb/tubesync#167]
I've had TubeSYnc setup for ~ 2 weeks now, and over the past 6 days random thumbnails have been downloaded with black bars added to the top/bottom of the image.

The thumbnail image between the black bars is correct, viewing the videos on YouTube shows the correct thumbnail resolution without black bars added, but viewing the Media in TubeSync shows a thumbnail with black bars.

If I re-download the video manually using youtube-dl, the correct thumbnail gets downloaded.

**Comment by meeb** [comment]
This might actually be `yt-dlp` doing thumbnails "properly" since the switch-over about a week ago. Could you try `yt-dlp` with `--add-metadata` from the command line and see if you still get black bars? If so this might be an upstream issue.

**Comment by ngInit** [comment]
Alright, so `yt-dlp https://www.youtube.com/watch?v=tNXDceetsiw --embed-metadata` doesn't actually save a thumbnail, however if I do `yt-dlp https://www.youtube.com/watch?v=tNXDceetsiw --write-thumbnail` the thumbnail has black bars. The same command ran through youtube-dl saves the thumbnail correctly.

**Comment by meeb** [comment]
`youtube-dl` doesn't support embedding thumbnails at all into MKV containers so what this probably is is "not working" vs "working" after the switch to `yt-dlp`. Either way, this very much looks like an upstream issue rather than anything I can do to fix it in TubeSync.

**Comment by ngInit** [comment]
Mmkay, so why is yt-dlp a requirement? If their repo is causing bugs, maybe it would be good to give your users an option of youtube-dl vs yt-dlp, stating certain yt-dlp features that aren't supported through youtube-dl might be disabled.

**Comment by meeb** [comment]
It's not a bug. Embedding thumbnails was broken in `youtube-dl` and it's actually working in `yt-dlp`.

Do you mean thumbnails in the actual TubeSync interface? Or for the media in a media server? Or both?

**Comment by ngInit** [comment]
I'm not talking about embedded thumbnails. I'm talking about the jpg thumbnails saved alongside the file.

I ran the same command with the `--write-thumbnail` flag on youtube-dl and yt-dlp. Youtube-dl saves the thumbnail image correctly, yt-dlp is incorrectly adding black bars. If there's no config option on your end to stop that, then I'd consider that a bug on yt-dlp.

**Comment by meeb** [comment]
Ah right. TubeSync just fetches whatever YouTube reports as the thumbnail. What's a URL to a video that has black bars on the thumbnail for you?

**Comment by ngInit** [comment]
Well, quite a few do. But it's inconsistent with a specific channel. The video listed in this comment had it earlier. I just tested it again now and it's still doing it.

> Alright, so `yt-dlp https://www.youtube.com/watch?v=tNXDceetsiw --embed-metadata` doesn't actually save a thumbnail, however if I do `yt-dlp https://www.youtube.com/watch?v=tNXDceetsiw --write-thumbnail` the thumbnail has black bars. The same command ran through youtube-dl saves the thumbnail correctly.

**Comment by ngInit** [comment]
The TubeSync WebUI also shows the thumbnail with black bars, presumably because it's using yt-dlp to fetch the thumbnail

**Comment by meeb** [comment]
That video returns thumbnails with black bars from YouTube, like

https://i.ytimg.com/vi/tNXDceetsiw/sddefault.jpg (height of 480px)

as the largest thumbnail. TubeSync fetches the video metadata using `yt-dlp` then downloads the thumbnail itself, using the largest available thumbnail before resizing it and storing it. The thumbnail I assume you would prefer is probably this one:

https://i.ytimg.com/vi/tNXDceetsiw/hqdefault.jpg?sqp=-oaymwEcCNACELwBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLCAjZUwY67ntQ1y3v7ymVP1LpmQsw (height of 188px)

`yt-dlp` returns substantially more information and can "impersonate" a mobile device which works around some increasingly common issues (region, login and age restrictions for example) that were becoming a big problem with vanilla `youtube-dl`.

What I imagine is going on is these thumbnails are probably meant for the mobile YouTube client. This might be annoying to resolve as the largest non-black-barred thumbnail on some videos is now smaller than the resolution used by TubeSync so the image would look pretty bad. Probably not worse than having black bars though.

This also seems to vary between YouTube returning a JPEG or WEBP image for the thumbnail so there's no immediately obvious generic rule to apply without further digging into it on how to decide which image to use.

I'll have a look further but this may not be a straight forward fix.

**Comment by ngInit** [comment]
Hmmm... Dunno how easy this would be compared to what you might be thinking now, but maybe there's some simple image post-processing script that can crop out the bars.

We could probably assume the image will always be in the center, so tell it to remove however many px rows from the top and bottom needed to make the image x, y scale consistent with whatever it should be for thumbnails. And hopefully without quality loss if it's editing jpgs.

**Comment by ngInit** [comment]
Oh.. maybe less convoluted of a solution - with `yt-dlp` impersonating mobile devices, is there a way to make it not impersonate a mobile device when it's downloading the thumbnails?

Maybe run a separate command after the download is complete with the `--write-thumbnail` flag tossed on there, it see's the video exists, skips the download, downloads the desktop/web version of the thumbnail.

**Comment by meeb** [comment]
Not all the images contain black bars or appear to be squashed to the same degree. I'm unlikely to code fuzzy logic that attempts to detect black bars and image distortions in an attempt to counter whatever YouTube is currently doing. Someone else might want to code that as a PR but I won't have time.

Realistically the only sensible answer I'd implement is to reliably find a way to get the non-distorted non-black-barred thumb of suitable size from YouTube in the first place.

This isn't the first time YouTube have changed stuff that made these upstream libraries return slightly undesirable information. It'll get fixed at some point probably, it just won't likely be that quick. I'd check upstream with `yt-dlp` to see if the black bars are reported there as well as that would be a much easier place to fix it than in TubeSync. This is mostly an upstream issue.

There doesn't appear to be an obvious API toggle to change `yt-dlp`'s behaviour in which scraping engine it selects for YouTube, but I've only given it a cursory glance so far.

TubeSync doesn't execute `yt-dlp` on the command line, it uses the embedded API from here:

https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/YoutubeDL.py

**Comment by ngInit** [comment]
Yeah lol - I wouldn't have figured that first solution would be likely. But ya gotta start brainstorming somewhere.

**Comment by ngInit** [comment]
Hey @meeb - I've been messing around with the command line version of youtube-dl and youtube-dlp these past few months. Something I noticed was whenever thumbnails were downloaded as a WEBP file instead of a jpg, I never ran into the problem of the black bar thumbnails.

I'm kind of going off memory, because it's been about a month and a half since I've had downloads setup and working, but I'm pretty sure the WEBP version either completely solved the black bar problem, or still included them but didn't squish the thumbnail between the black bars to fit the standard thumbnail resolution.

Obviously there's the problem that WEBP isn't really supported by a lot of media servers, but at least in my case it was easy enough to write a script to use ffmpeg to auto convert all WEBP files.

An option to choose what version of the thumbnail to save would be ideal, webp or jpg I'm guessing are the two main ones.

**Comment by meeb** [comment]
The thumbnails are generally already downloaded as WEBP and converted to JPEG. The source WEBP images have the black bars on. `yt-dlp` impersonates a mobile client in many requests and it seems YouTube response with black barred images to mobile devices. Some of them do indeed appear to be squashed normal images and stretching the image out over the black bars works occasionally, but it's not something that I can see that has an easily identifiable pattern to automate. Feel free to paste examples of commands that work for you, I may well have missed something or something may have changed.

**Comment by ngInit** [comment]
Well as I said, I haven't had a YouTube downloader running in a few months since I've been reassembling my server. Next time I notice it I'll do some tests on that video and let you know

**Comment by ngInit** [comment]
Well that was pretty fast. Alright so for this video https://www.youtube.com/watch?v=nvIPX-9CpS0

If you run the command `yt-dlp --list-thumbnails https://www.youtube.com/watch?v=nvIPX-9CpS0`

outputs:
```
[youtube] nvIPX-9CpS0: Downloading webpage
[youtube] nvIPX-9CpS0: Downloading android player API JSON
[info] Available thumbnails for nvIPX-9CpS0:
ID Width   Height  URL
0  unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/3.jpg
1  unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/3.webp
2  unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/2.jpg
3  unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/2.webp
4  unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/1.jpg
5  unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/1.webp
6  120     90      https://i.ytimg.com/vi/nvIPX-9CpS0/default.jpg
7  unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/default.webp
8  unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/mq3.jpg
9  unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/mq3.webp
10 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/mq2.jpg
11 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/mq2.webp
12 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/mq1.jpg
13 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/mq1.webp
14 320     180     https://i.ytimg.com/vi/nvIPX-9CpS0/mqdefault.jpg
15 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/mqdefault.webp
16 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/0.jpg
17 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/0.webp
18 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/hq3.jpg
19 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/hq3.webp
20 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/hq2.jpg
21 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/hq2.webp
22 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/hq1.jpg
23 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/hq1.webp
24 168     94      https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEbCKgBEF5IVfKriqkDDggBFQAAiEIYAXABwAEG&rs=AOn4CLDgCRBozZydfYgKIboUWc1pDk7UlQ
25 196     110     https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEbCMQBEG5IVfKriqkDDggBFQAAiEIYAXABwAEG&rs=AOn4CLBFZj61Jc5oXgxgX-K53jKY07ovXg
26 246     138     https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEcCPYBEIoBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLBmdeH9lNvJ0LcJs7NUzhQRqzfarQ
27 336     188     https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEcCNACELwBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLBizgKHU5W5snJHB-MvqHpJJbu4sQ
28 480     360     https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg
29 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/hqdefault.webp
30 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/sd3.jpg
31 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/sd3.webp
32 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/sd2.jpg
33 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/sd2.webp
34 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/sd1.jpg
35 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/sd1.webp
36 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/sddefault.jpg
37 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/sddefault.webp
38 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/hq720.jpg
39 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/hq720.webp
40 unknown unknown https://i.ytimg.com/vi/nvIPX-9CpS0/maxresdefault.jpg
41 unknown unknown https://i.ytimg.com/vi_webp/nvIPX-9CpS0/maxresdefault.webp
```

These thumbnails all have black bars, however, they are the correct dimensions are are displayed correctly when they're downloaded:
```
https://i.ytimg.com/vi/nvIPX-9CpS0/default.jpg
https://i.ytimg.com/vi_webp/nvIPX-9CpS0/default.webp
https://i.ytimg.com/vi/nvIPX-9CpS0/mqdefault.jpg
https://i.ytimg.com/vi_webp/nvIPX-9CpS0/mqdefault.webp
https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEbCKgBEF5IVfKriqkDDggBFQAAiEIYAXABwAEG&rs=AOn4CLDgCRBozZydfYgKIboUWc1pDk7UlQ
https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEbCMQBEG5IVfKriqkDDggBFQAAiEIYAXABwAEG&rs=AOn4CLBFZj61Jc5oXgxgX-K53jKY07ovXg
https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEcCPYBEIoBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLBmdeH9lNvJ0LcJs7NUzhQRqzfarQ
https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg?sqp=-oaymwEcCNACELwBSFXyq4qpAw4IARUAAIhCGAFwAcABBg==&rs=AOn4CLBizgKHU5W5snJHB-MvqHpJJbu4sQ
https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg
https://i.ytimg.com/vi_webp/nvIPX-9CpS0/hqdefault.webp
```

From what I can tell, TubeSync is changing the dimensions of the thumbnail when it saves/converts it from a webp, so whenever there are black bars the output gets squished like this
![image](https://user-images.githubusercontent.com/38122413/198371212-024ad122-f37b-4ece-ac61-e6a3e46b4f3d.png)

**Comment by ngInit** [comment]
I found these lines in tasks.py
```
width = getattr(settings, 'MEDIA_THUMBNAIL_WIDTH', 430)
    height = getattr(settings, 'MEDIA_THUMBNAIL_HEIGHT', 240)
    i = get_remote_image(url)
    log.info(f'Resizing {i.width}x{i.height} thumbnail to '
             f'{width}x{height}: {url}')
    i = resize_image_to_height(i, width, height)
```

I'm wondering if this could be fixed by adding a check to determines if black bars exist or not, and if they exist set `MEDIA_THUMBNAIL_WIDTH` to 480 and `MEDIA_THUMBNAIL_HEIGHT` to 360 instead of 430, 240.

**Comment by meeb** [comment]
Hm, I could use different dimensions for the thumbnails, however this would look pretty wrong in the web interface which expects 1:1.33 style sizes if I remember correctly. Currently TubeSync just uses whatever `yt-dlp` provides as the `thumbnail` value in the metadata as the source. There are loads of thumbnail options but I still can't see a clear sensible way to detect black bars, other than I guess actually just checking if the first and last rows are all black or something.

**Comment by ngInit** [comment]
Yeah, so I changed those resize variables and the thumbnails in the webUI was the correct dimensions, but it screwed up the CSS. Something along these lines will fix that issue:

![image](https://user-images.githubusercontent.com/38122413/198840084-8b149676-d04d-4b06-85a0-1b2b092d507d.png)
```
{
    background-image: url(https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg);
    height: 105px;
    background-position: center;
    background-size: cover;
    width: 188px;
}
```

![image](https://user-images.githubusercontent.com/38122413/198840248-be4d1dd4-cc0c-4126-a82b-b851adcb178e.png)
```
 {
    background-image: url(https://i.ytimg.com/vi/nvIPX-9CpS0/hqdefault.jpg);
    height: 190px;
    background-position: center;
    background-size: cover;
    width: 340px;
}
```

You'd obviously want to change the hard coded width or height out to make it work with the responsive web design, but proof of concept the background-position center and background-size cover should fix that.

________________________

And then as far as detecting the black bar image, my first idea would be to grab the dimensions of the image and do something like this:
```
if (IMAGE_X_DIMENSION / IMAGE_Y_DIMENSION > 1.34) {
  // normal resize
} else {
  // black bars resize
}
```

**Comment by tcely** [comment]
I don't remember where I read this, it was likely in a Google API document, but the bars are added to the original image when YouTube post processing determines the uploaded thumbnail isn't the correct aspect ratio.

I wouldn't expect the format of the thumbnail to make any difference. Also, we would need to do more than look at the image dimensions after downloading, because it should be a standard size like any other thumbnail.
